tfjs-backend-wasm/src/cc/kernels/Conv2D.cc

```diff
@@ -0,0 +1,144 @@
+std::map<operator_cache_key, xnn_operator_t> operator_cache;
+
+// Maps a filter id to a list of operator cache keys that this filter belongs
+// to.
+std::unordered_map<int, std::vector<operator_cache_key>>
+    filter_operator_cache_key_map;
+
+void delete_xnn_operators(int filter_id) {
+  std::vector<operator_cache_key> operator_cache_keys =
+      filter_operator_cache_key_map.at(filter_id);
+  for (auto& operator_cache_key : operator_cache_keys) {
+    auto conv2d_op = operator_cache.at(operator_cache_key);
+    xnn_delete_operator(conv2d_op);
+    tfjs::backend::xnn_operator_count--;
+    operator_cache.erase(operator_cache_key);
+  }
+}
+}  // namespace
+
+namespace tfjs {
+namespace wasm {
+// We use C-style API to interface with Javascript.
+extern "C" {
+
+#ifdef __EMSCRIPTEN__
+EMSCRIPTEN_KEEPALIVE
+#endif
+void Conv2D(int x_id, int batch_size, int input_height, int input_width,
+            int filter_id, int filter_height, int filter_width, int pad_top,
+            int pad_right, int pad_bottom, int pad_left, int dilation_height,
+            int dilation_width, int stride_height, int stride_width,
+            int input_channels, int output_channels, int out_id) {
+  auto& x_info = backend::get_tensor_info(x_id);
+  auto& filter_info = backend::get_tensor_info(filter_id);
+  auto& out_info = backend::get_tensor_info(out_id);
+
+  const float* x_buf = reinterpret_cast<float*>(x_info.memory_offset);
+  const float* filter_buf = reinterpret_cast<float*>(filter_info.memory_offset);
+  float* out_buf = reinterpret_cast<float*>(out_info.memory_offset);
+
+  xnn_operator_t conv2d_op = nullptr;
+
+  const int flags = 0;
+  const int groups = 1;
+
+  operator_cache_key cache_key = {
+      pad_top,         pad_right,      pad_bottom,    pad_left,
+      filter_height,   filter_width,   stride_height, stride_width,
+      dilation_height, dilation_width, groups,        input_channels,
+      output_channels, filter_id,      flags};
+
+  auto operator_cache_idx = operator_cache.find(cache_key);
+  if (operator_cache_idx == operator_cache.end()) {
+    float output_min = -std::numeric_limits<float>::infinity();
+    float output_max = std::numeric_limits<float>::infinity();
+
+    const float* bias_buf = nullptr;
+    xnn_status status = xnn_create_convolution2d_nhwc_f32(
+        pad_top, pad_right, pad_bottom, pad_left, filter_height, filter_width,
+        stride_height, stride_width, dilation_height, dilation_width, groups,
+        input_channels /* group_input_channels */,
+        output_channels /* group_output_channels */,
+        input_channels /* input_pixel_stride */,
+        output_channels /* output_pixel_stride */, filter_buf, bias_buf,
+        output_min, output_max, flags, &conv2d_op);
+    if (status != xnn_status_success) {
+      util::warn(
+          "XNN status for xnn_create_convolution2d_nhwc_f32 is not successful. "
+          "Got status %d. Use -c dbg to see XNN logs.",
+          status);
```

Comment:
does `util::warn` throw an exception, or otherwise exit the function? if operation creation failed, it makes no sense to continue